gammapy/catalog/core.py

```diff
@@ -257,18 +269,50 @@ def _make_source_object(self, index):
             name_extended = data["Source_Name"].strip()
         else:
             name_extended = None
+
+        name_spectral = self._get_name_spectral(data)
+
         try:
             idx = self._lookup_extended_source_idx[name_extended]
             data_extended = table_row_to_dict(self.extended_sources_table[idx])
         except (KeyError, AttributeError):
             data_extended = None
 
-        source = self.source_object_class(data, data_extended)
+        try:
+            idx = self._lookup_spectral_source_idx[name_spectral]
+            data_spectral = table_row_to_dict(self.spectral_table[idx])
+        except (KeyError, AttributeError):
+            data_spectral = None
+
+        source = self.source_object_class(data, data_extended, data_spectral)
         return source
 
+    def _get_name_spectral(self, data):
+        if "Source_name" in data:
+            name_spectral = data["Source_name"].strip()
+        else:
+            name_spectral = None
+        return name_spectral
+
+    def _get_spectral_table_source_name_key(self):
+        return self._source_name_key
+
+    @lazyproperty
+    def _lookup_spectral_source_idx(self):
+        """Return a dictionnary of names-idx pairs corresponding to the
+        entry of the spectral table (for Fermi-LAT pulsar catalogs).
+        """
+        source_name_key = self._get_spectral_table_source_name_key()
+        names = [_.strip() for _ in self.spectral_table[source_name_key]]
+        idx = range(len(names))
+        return dict(zip(names, idx))
```

Comment:
The code is very similar between `_lookup_spectral_source_idx` and `_lookup_extended_source_idx`. It could be turned a single function taking either `self.spectral_table[source_name_key]` or `self.extended_sources_table[self._source_name_key]`as argument